### PR TITLE
[C++ Worker]Add some api of placement group

### PR DESCRIPTION
**Pull request by qicosmos**
## Why are these changes needed?

Set placement group when create an actor or call a ray task.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests

### CODE REVIEW

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 1]
```diff
@@ -276,17 +276,44 @@ TEST(RayClusterModeTest, GetActorTest) {
...
+  auto first_placement_group = CreateSimplePlacementGroup("first_placement_group");
   ray::RemovePlacementGroup(first_placement_group.GetID());
 }
 
+TEST(RayClusterModeTest, CreateActorWithPlacementGroup) {
+  auto placement_group = CreateSimplePlacementGroup("first_placement_group");
+  EXPECT_TRUE(ray::WaitPlacementGroupReady(placement_group.GetID(), 10));
+
+  auto actor1 = ray::Actor(RAY_FUNC(Counter::FactoryCreate))
+                    .SetResources({{"CPU", 1.0}})
```
Comment (SongGuyang) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:297]: If the required resources exceeds the resources of `PlacementGroup`, what will be happened? Can you add this test case?

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 2]
```diff
@@ -276,17 +276,44 @@ TEST(RayClusterModeTest, GetActorTest) {
...
+}
 
+TEST(RayClusterModeTest, CreateAndRemovePlacementGroup) {
+  auto first_placement_group = CreateSimplePlacementGroup("first_placement_group");
   ray::RemovePlacementGroup(first_placement_group.GetID());
 }
 
+TEST(RayClusterModeTest, CreateActorWithPlacementGroup) {
+  auto placement_group = CreateSimplePlacementGroup("first_placement_group");
+  EXPECT_TRUE(ray::WaitPlacementGroupReady(placement_group.GetID(), 10));
```
Comment (SongGuyang) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:294]: If the resources of PlacementGroup  exceeds the resources of Cluster, what will be happened? Can you add this test case?

File: cpp/src/ray/runtime/task/native_task_submitter.cc [hunk 3]
```diff
@@ -30,6 +30,15 @@ RayFunction BuildRayFunction(InvocationSpec &invocation) {
   return RayFunction(ray::Language::CPP, function_descriptor);
 }
 
+auto GetBundleID = [](const auto &options) {
```
Comment (SongGuyang) [inline cpp/src/ray/runtime/task/native_task_submitter.cc:33]: Can you make this lambda to be a static function?

**Review by SongGuyang — COMMENTED** [review]
Can we add more tests?

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 4]
```diff
@@ -276,17 +276,67 @@ TEST(RayClusterModeTest, GetActorTest) {
...
+
+TEST(RayClusterModeTest, CreatePlacementGroupExceedsClusterResource) {
+  std::vector<std::unordered_map<std::string, double>> bundles{{{"CPU", 100}}};
+
   ray::internal::PlacementGroupCreationOptions options{
       false, "first_placement_group", bundles, ray::internal::PlacementStrategy::PACK};
   auto first_placement_group = ray::CreatePlacementGroup(options);
-  EXPECT_TRUE(ray::WaitPlacementGroupReady(first_placement_group.GetID(), 10));
-
+  EXPECT_FALSE(ray::WaitPlacementGroupReady(first_placement_group.GetID(), 3));
```
Comment (SongGuyang) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:300]: 3s or 3ms?

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 5]
```diff
@@ -276,17 +276,67 @@ TEST(RayClusterModeTest, GetActorTest) {
...
+
+TEST(RayClusterModeTest, CreateAndRemovePlacementGroup) {
+  auto first_placement_group = CreateSimplePlacementGroup("first_placement_group");
+  EXPECT_THROW(CreateSimplePlacementGroup("first_placement_group"),
+               ray::internal::RayException);
+  ray::RemovePlacementGroup(first_placement_group.GetID());
+}
+
+TEST(RayClusterModeTest, CreatePlacementGroupExceedsClusterResource) {
+  std::vector<std::unordered_map<std::string, double>> bundles{{{"CPU", 100}}};
```
Comment (SongGuyang) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:295]: Maybe we should give a bigger cpu number? I concern this test run in a big machine.

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 6]
```diff
@@ -276,17 +276,67 @@ TEST(RayClusterModeTest, GetActorTest) {
...
+TEST(RayClusterModeTest, CreateActorWithPlacementGroup) {
+  auto placement_group = CreateSimplePlacementGroup("first_placement_group");
+  EXPECT_TRUE(ray::WaitPlacementGroupReady(placement_group.GetID(), 10));
+
+  auto actor1 = ray::Actor(RAY_FUNC(Counter::FactoryCreate))
+                    .SetResources({{"CPU", 1.0}})
+                    .SetPlacementGroup(placement_group, 0)
+                    .Remote();
+  auto r1 = actor1.Task(&Counter::Plus1).Remote();
+  EXPECT_EQ(*r1.Get(), 1);
```
Comment (SongGuyang) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:313]: We should also use ray::Wait here to compare with the exceeding case?

File: cpp/src/ray/test/cluster/cluster_mode_test.cc [hunk 7]
```diff
@@ -276,17 +276,67 @@ TEST(RayClusterModeTest, GetActorTest) {
...
+  ray::internal::PlacementGroupCreationOptions options{
+      false, name, bundles, ray::internal::PlacementStrategy::PACK};
+  return ray::CreatePlacementGroup(options);
+}
+
+TEST(RayClusterModeTest, CreateAndRemovePlacementGroup) {
+  auto first_placement_group = CreateSimplePlacementGroup("first_placement_group");
+  EXPECT_THROW(CreateSimplePlacementGroup("first_placement_group"),
+               ray::internal::RayException);
+  ray::RemovePlacementGroup(first_placement_group.GetID());
```
Comment (SongGuyang) [inline cpp/src/ray/test/cluster/cluster_mode_test.cc:291]: Can we also use `ray::WaitPlacementGroupReady` to check result here?

Comment (qicosmos) [reply cpp/src/ray/test/cluster/cluster_mode_test.cc:291]: If creating a same name placement group, the coreworker will return a non-ok status, and then throw an exception, there is no pd_id returned, so can't use WaitPlacementGroupReady(pg_id) here.

**Review by SongGuyang — APPROVED** [review]
LGTM

**Review by raulchen — APPROVED** [review]

**Pull request merged**